File: android/src/main/kotlin/com/enver/lightweight_barcode_scanner/ScannerConfiguration.kt
package com.enver.lightweight_barcode_scanner

/** Analysis resolution, as a plain pair so this file stays framework-free. */
data class AnalysisResolution(val width: Int, val height: Int)

/** Region of interest as fractions (0..1) of the upright frame. */
data class ScanArea(
    val left: Float,
    val top: Float,
    val width: Float,
    val height: Float,
)

/**
 * Scanner settings as they arrive from Dart.
 *
 * Deliberately free of Android types: this is the mirror image of
 * `ScannerOptions.toMap()` on the Dart side and the place where the two are
 * most likely to drift apart, so it has to be unit-testable on the JVM.
 */
data class ScannerConfiguration(
    val formats: Int,
    val scanMode: String,
    val facing: String,
    val resolution: AnalysisResolution,
    val profile: String,
    val duplicateFilterMillis: Long,
    val detectionsPerSecond: Int,
    val includeRawBytes: Boolean,
    val torchEnabled: Boolean,
    val scanRegion: ScanArea?,
) {
    val isFrontFacing: Boolean get() = facing == "front"

    val maxSymbols: Int get() = if (scanMode == "multiple") 16 else 1

    val stopAfterFirstResult: Boolean get() = scanMode == "single"

    /** Minimum gap between decode attempts. Surplus frames are dropped. */
    val minAnalysisIntervalMillis: Long get() = 1000L / detectionsPerSecond

    val tryHarder: Boolean get() = profile != "fast"
    val tryRotate: Boolean get() = profile != "fast"
    val tryInvert: Boolean get() = profile == "thorough"
    val tryDownscale: Boolean get() = profile != "fast"

    companion object {
        fun fromMap(arguments: Map<*, *>): ScannerConfiguration = ScannerConfiguration(
            formats = (arguments["formats"] as? Number)?.toInt() ?: 0,
            scanMode = arguments["scanMode"] as? String ?: "continuous",
            facing = arguments["facing"] as? String ?: "back",
            resolution = resolutionFor(arguments["resolution"] as? String),
            profile = arguments["profile"] as? String ?: "balanced",
            duplicateFilterMillis =
                (arguments["duplicateFilterMillis"] as? Number)?.toLong() ?: 750L,
            detectionsPerSecond =
                ((arguments["detectionsPerSecond"] as? Number)?.toInt() ?: 12)
                    .coerceIn(1, 60),
            includeRawBytes = arguments["includeRawBytes"] as? Boolean ?: false,
            torchEnabled = arguments["torchEnabled"] as? Boolean ?: false,
            scanRegion = scanAreaFrom(arguments["scanRegion"]),
        )

        /** The wire format is `[left, top, width, height]`, all fractions. */
        fun scanAreaFrom(value: Any?): ScanArea? {
            val values = value as? List<*> ?: return null
            if (values.size < 4) return null
            return ScanArea(
                (values[0] as Number).toFloat(),
                (values[1] as Number).toFloat(),
                (values[2] as Number).toFloat(),
                (values[3] as Number).toFloat(),
            )
        }

        private fun resolutionFor(name: String?): AnalysisResolution = when (name) {
            "low" -> AnalysisResolution(640, 480)
            "high" -> AnalysisResolution(1920, 1080)
            else -> AnalysisResolution(1280, 720)
        }
    }
}

File: ios/lightweight_barcode_scanner/Sources/lightweight_barcode_scanner/ScannerConfiguration.swift
import AVFoundation
import Foundation

/// Scanner settings as they arrive from Dart.
struct ScannerConfiguration {
  var formats: UInt32
  var scanMode: String
  var facing: String
  var resolution: AVCaptureSession.Preset
  var profile: String
  var duplicateFilterMillis: Double
  var detectionsPerSecond: Int
  var includeRawBytes: Bool
  var torchEnabled: Bool
  var scanRegion: CGRect?

  var isFrontFacing: Bool { facing == "front" }
  var maxSymbols: Int { scanMode == "multiple" ? 16 : 1 }
  var stopAfterFirstResult: Bool { scanMode == "single" }

  /// Minimum gap between decode attempts. Surplus frames are dropped.
  var minAnalysisInterval: CFTimeInterval { 1.0 / Double(detectionsPerSecond) }

  var tryHarder: Bool { profile != "fast" }
  var tryRotate: Bool { profile != "fast" }
  var tryInvert: Bool { profile == "thorough" }
  var tryDownscale: Bool { profile != "fast" }

  static func from(_ arguments: [String: Any]) -> ScannerConfiguration {
    var region: CGRect?
    if let values = arguments["scanRegion"] as? [NSNumber], values.count >= 4 {
      region = CGRect(
        x: values[0].doubleValue,
        y: values[1].doubleValue,
        width: values[2].doubleValue,
        height: values[3].doubleValue
      )
    }
    return ScannerConfiguration(
      formats: (arguments["formats"] as? NSNumber)?.uint32Value ?? 0,
      scanMode: arguments["scanMode"] as? String ?? "continuous",
      facing: arguments["facing"] as? String ?? "back",
      resolution: preset(for: arguments["resolution"] as? String),
      profile: arguments["profile"] as? String ?? "balanced",
      duplicateFilterMillis:
        (arguments["duplicateFilterMillis"] as? NSNumber)?.doubleValue ?? 750,
      detectionsPerSecond: min(
        60, max(1, (arguments["detectionsPerSecond"] as? NSNumber)?.intValue ?? 12)),
      includeRawBytes: arguments["includeRawBytes"] as? Bool ?? false,
      torchEnabled: arguments["torchEnabled"] as? Bool ?? false,
      scanRegion: region
    )
  }

  private static func preset(for name: String?) -> AVCaptureSession.Preset {
    switch name {
    case "low": return .vga640x480
    case "high": return .hd1920x1080
    default: return .hd1280x720
    }
  }
}
